Add support for gssapi-with-mic

Qalthos opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #153 

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

Qalthos committed:
Add support for gssapi-with-mic

Qalthos committed:
Treat gssapi_delegate_credentials as a number

Qalthos committed:
Whoops, wrong gssapi method

ganeshrn commented on src/pylibsshext/session.pyx:
```diff
@@ -146,6 +149,7 @@ cdef class Session(object):
     def set_ssh_options(self, key, value):
         cdef unsigned int port_i
```
Can we use generic variable names here. Since `set_ssh_options` is called for each option in loop the cdef variables are used to store the value to apporicate datatype. How about `value_int`, `value_long_int` `value_unsigned_int`?

Qalthos replied:
Yeah, that was what I was thinking would be a good path for these variables.

ganeshrn commented on src/pylibsshext/session.pyx:
```diff
@@ -32,6 +32,9 @@ OPTS_MAP = {
     "timeout": libssh.SSH_OPTIONS_TIMEOUT,
     "knownhosts": libssh.SSH_OPTIONS_KNOWNHOSTS,
     "proxycommand": libssh.SSH_OPTIONS_PROXYCOMMAND,
+    "gssapi_server_identity": libssh.SSH_OPTIONS_GSSAPI_SERVER_IDENTITY,
```
Update the `connect` API doc to reflect the newly added gssapi options

webknjaz commented on src/pylibsshext/session.pyx:
```diff
@@ -444,6 +460,20 @@ cdef class Session(object):
+        rc = libssh.ssh_userauth_gssapi(self._libssh_session)
+
+        if rc in (libssh.SSH_AUTH_ERROR, libssh.SSH_AUTH_DENIED):
+            raise LibsshSessionException("Failed to authenticate with keyboard-interactive: {err}".format(err=self._get_session_error_str()))
```
The error message wasn't updated:
```suggestion
            raise LibsshSessionException("Failed to authenticate via gssapi-with-mic: {err}".format(err=self._get_session_error_str()))
```

Qalthos committed:
Fix error message to report the correct authentication type

Qalthos committed:
reduce specificity of typed C variables in set_ssh_options

webknjaz commented:
@Qalthos mind rebasing?

Qalthos committed:
Add example gssapi to documentation

Qalthos committed:
Add GSSAPI params to connect() docstring

webknjaz commented on docs/_samples/gssapi.py:
```diff
@@ -0,0 +1,32 @@
+print(f'{ssh.is_connected=}')
+if ssh.is_connected:
+    chan_shell = ssh.invoke_shell()
+    data_b = chan_shell.read_bulk_response(timeout=2, retry=10)
+    chan_shell.close()
```
I think that when showcasing any code with closable resources, it's best to advertise practices of calling `.close()` in a robust manner like
```suggestion
    try:
        data_b = chan_shell.read_bulk_response(timeout=2, retry=10)
    finally:
        chan_shell.close()
        ssh.close()
        raise SystemExit
```
@Qalthos WDYT?

webknjaz commented on docs/user_guide.rst:
```diff
@@ -48,6 +48,22 @@ Connecting with remote SSH server
+
+.. literalinclude:: _samples/gssapi.py
+   :language: python
+   :lines: 11-21
```
It may be better maintainable to use markers for such simple examples:
```suggestion
   :start-after: try:
   :end-before: except LibsshSessionException as ssh_exc:
   :dedent: 4
   :emphasize-lines: 8-9
```
(I haven't tried this locally, so better test first)

Qalthos committed:
Update user guide and examples

User guide includes from examples now pinned to specific text in
samples, instead of line numbers

ganeshrn reviewed (approved):
Looks good. Thank you!

webknjaz commented:
@Qalthos here's the testing tutorial https://www.confluent.io/blog/containerized-testing-with-kerberos-and-ssh/ that @ganeshrn mentioned earlier today (posting for history).

If you don't figure out the integration test part, feel free to merge this as is but create an issue task for exploring this later.
